store/sqlstore/migration.go

```diff
@@ -0,0 +1,194 @@
+	AsyncMigrationStatusRun      AsyncMigrationStatus = "run"      // migration should be run
+	AsyncMigrationStatusSkip     AsyncMigrationStatus = "skip"     // migration should be skipped (not sure if needed?)
+	AsyncMigrationStatusComplete AsyncMigrationStatus = "complete" // migration was already executed
+	AsyncMigrationStatusFailed   AsyncMigrationStatus = "failed"   // migration has failed
+)
+
+// AsyncMigration executes a single database migration that allows concurrent DML
+type AsyncMigration interface {
+	// name of migration, must be unique among migrations - used for saving status in database
+	Name() string
+	// returns if migration should be run / was already executed
+	GetStatus(SqlStore) (AsyncMigrationStatus, error)
+	// exectutes migration, gets started transaction with lock timeouts set
+	Execute(context.Context, SqlStore, *sql.Conn) (AsyncMigrationStatus, error)
+}
+
+// MigrationRunner runs queued async migrations
+type MigrationRunner struct {
+	supplier   SqlStore
+	migrations []AsyncMigration
+	done       chan struct{}
+	options    MigrationOptions
+	cancelFn   context.CancelFunc
+}
+
+type MigrationOptions struct {
+	LockTimeout int // Lock timeout in seconds
+	BackoffTime int // Time to wait between retries
+	NumRetries  int // Number of retries in case of failed migration
+}
+
+func NewMigrationRunner(s SqlStore, o MigrationOptions) *MigrationRunner {
+	if o.LockTimeout == 0 {
+		o.LockTimeout = MigrationDefaultLockTimeout
+	}
+	if o.BackoffTime == 0 {
+		o.BackoffTime = MigrationDefaultBackoffTime
+	}
+	if o.NumRetries == 0 {
+		o.NumRetries = MigrationDefaultNumRetries
+	}
+	return &MigrationRunner{
+		supplier: s,
+		done:     make(chan struct{}),
+		options:  o,
+	}
+}
+
+// Add checks if the migration should be executed and adds it to queue
+func (r *MigrationRunner) Add(m AsyncMigration) error {
+	// check status in Systems table
+	currentStatus, appErr := r.supplier.System().GetByName("migration_" + m.Name())
+	if appErr == nil {
+		if currentStatus.Value == string(AsyncMigrationStatusComplete) || currentStatus.Value == string(AsyncMigrationStatusSkip) {
+			return nil
+		}
+	}
+	// get status from migration
+	status, err := m.GetStatus(r.supplier)
+	if err != nil {
+		return err
+	}
+	if status == AsyncMigrationStatusComplete || status == AsyncMigrationStatusSkip {
+		return nil
+	}
+	r.migrations = append(r.migrations, m)
+	return nil
+}
+
+// Run all queued migrations sequentially
+func (r *MigrationRunner) Run() error {
+	go func() {
+		ctx, cancelFn := context.WithCancel(context.Background())
+		defer cancelFn()
+		defer close(r.done)
+		r.cancelFn = cancelFn
+
+		for _, m := range r.migrations {
+			// function that will try to execute migration
+			migrate := func() error {
+				var err error
+				conn, err := createConnectionWithLockTimeout(ctx, r.supplier, r.options.LockTimeout)
+				if err != nil {
+					mlog.Error("Failed to setup connection", mlog.Err(err))
```

Comment:
Let's do all the logging one level above at the call site. And just return wrapped errors from here.